File: FrontEnd/lib/data/mappers/plant_species_mapper.dart
/// @file plant_species_mapper.dart
/// @description Implementación del mapper de especies de plantas.
/// Convierte PlantSpeciesModel ↔ PlantSpecies normalizando tipos y valores.
/// @module Plants
/// @layer Data
library;

import '../../core/utils/date_utils.dart';
import '../../domain/entities/plant_species.dart';
import '../i_mappers/i_plant_species_mapper.dart';
import '../models/plant_species_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// PLANT SPECIES MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Implementación de [IPlantSpeciesMapper].
///
/// [implements] IPlantSpeciesMapper
/// [injectable] registrar en container.dart como singleton.
class PlantSpeciesMapper implements IPlantSpeciesMapper {

  // ─── PlantSpeciesModel → PlantSpecies ─────────────────────────────────────────

  @override
  PlantSpecies toEntity(PlantSpeciesModel model) {
    return PlantSpecies(
      id:                   model.id,
      name:                 model.name,
      scientificName:       model.scientificName,
      image:                model.image,
      climateCompatibility: List.unmodifiable(model.climateCompatibility),
      tips:                 List.unmodifiable(model.tips),
      isPublic:             model.isPublic,
      createdBy:            model.createdBy,
      createdAt:            PlantDateUtils.parseUtc(model.createdAt) ?? DateTime.now().toUtc(),
      updatedAt:            PlantDateUtils.parseUtc(model.updatedAt) ?? DateTime.now().toUtc(),
      careRequirements:     _toCareRequirements(model.careRequirements),
      requiresPruning:             model.requiresPruning,
      pruningMonths:               model.pruningMonths != null
          ? List.unmodifiable(model.pruningMonths!)
          : null,
      produceFruit:                model.produceFruit,
      harvestMonths:               model.harvestMonths != null
          ? List.unmodifiable(model.harvestMonths!)
          : null,
      seasonalWateringAdjustment:  model.seasonalWateringAdjustment != null
          ? SeasonalWateringAdjustment(
              summer: model.seasonalWateringAdjustment!.summer,
              winter: model.seasonalWateringAdjustment!.winter,
            )
          : null,
      minRainfallMm:               model.minRainfallMm,
      waterLitersPerWatering:      model.waterLitersPerWatering,
    );
  }

  // ─── PlantSpecies → PlantSpeciesModel ─────────────────────────────────────────

  @override
  PlantSpeciesModel toModel(PlantSpecies entity) {
    return PlantSpeciesModel(
      id:                   entity.id,
      name:                 entity.name,
      scientificName:       entity.scientificName,
      image:                entity.image,
      climateCompatibility: entity.climateCompatibility,
      tips:                 entity.tips,
      isPublic:             entity.isPublic,
      createdBy:            entity.createdBy,
      createdAt:            PlantDateUtils.toIso8601(entity.createdAt),
      updatedAt:            PlantDateUtils.toIso8601(entity.updatedAt),
      careRequirements:     _toCareRequirementsModel(entity.careRequirements),
      requiresPruning:             entity.requiresPruning,
      pruningMonths:               entity.pruningMonths,
      produceFruit:                entity.produceFruit,
      harvestMonths:               entity.harvestMonths,
      seasonalWateringAdjustment:  entity.seasonalWateringAdjustment != null
          ? SeasonalWateringAdjustmentModel(
              summer: entity.seasonalWateringAdjustment!.summer,
              winter: entity.seasonalWateringAdjustment!.winter,
            )
          : null,
      minRainfallMm:               entity.minRainfallMm,
      waterLitersPerWatering:      entity.waterLitersPerWatering,
    );
  }

  // ─── Helpers de careRequirements ──────────────────────────────────────────────

  SpeciesCareRequirements _toCareRequirements(SpeciesCareRequirementsModel model) {
    return SpeciesCareRequirements(
      wateringDays:     model.wateringDays,
      lightNeed:        model.lightNeed,
      temperatureRange: model.temperatureRange != null
          ? SpeciesTemperatureRange(
              min: model.temperatureRange!.min,
              max: model.temperatureRange!.max,
            )
          : null,
    );
  }

  SpeciesCareRequirementsModel _toCareRequirementsModel(SpeciesCareRequirements reqs) {
    return SpeciesCareRequirementsModel(
      wateringDays:     reqs.wateringDays,
      lightNeed:        reqs.lightNeed,
      temperatureRange: reqs.temperatureRange != null
          ? SpeciesTemperatureRangeModel(
              min: reqs.temperatureRange!.min,
              max: reqs.temperatureRange!.max,
            )
          : null,
    );
  }
}

File: BackEnd/src/data/mappers/plant_species_mapper.ts
/**
 * @file plant_species_mapper.ts
 * @description Implementación del mapper de especies de plantas.
 * Convierte entre PlantSpeciesDocument (MongoDB), PlantSpecies (dominio) y PlantSpeciesResponseDTO.
 * @module Plants
 * @layer Data
 *
 * @implements {IPlantSpeciesMapper}
 * @injectable
 */

import { injectable } from 'inversify';
import { ObjectId } from 'mongodb';
import type { IPlantSpeciesMapper } from '../IMappers/IPlantSpeciesMapper.js';
import { PlantSpecies, CareRequirements, AuditEntry } from '../../domain/entities/PlantSpecies.js';
import type { LightNeed } from '../../domain/entities/Plant.js';
import { PlantSpeciesDocument, CareRequirementsDocument, AuditEntryDocument } from '../datasources/mongodb/models/PlantSpeciesModel.js';
import type { PlantSpeciesResponseDTO } from '../../domain/dtos/plants/plant-species-response.dto.js';

/**
 * Mapper de especies de plantas.
 *
 * @implements {IPlantSpeciesMapper}
 * @injectable
 */
@injectable()
export class PlantSpeciesMapper implements IPlantSpeciesMapper {

  /**
   * Convierte un documento MongoDB a entidad PlantSpecies.
   *
   * @param doc — Documento de la colección 'plant_species'.
   * @returns Entidad PlantSpecies.
   */
  toEntity(doc: PlantSpeciesDocument): PlantSpecies {
    return new PlantSpecies({
      id:                   doc._id.toHexString(),
      name:                 doc.name,
      scientificName:       doc.scientificName,
      image:                doc.image,
      careRequirements:     this._toCareRequirements(doc.careRequirements),
      climateCompatibility: doc.climateCompatibility,
      tips:                 doc.tips,
      createdBy:            doc.createdBy?.toHexString(),
      isPublic:             doc.isPublic,
      requiresPruning:      doc.requiresPruning,
      pruningMonths:         doc.pruningMonths,
      produceFruit:                 doc.produceFruit,
      harvestMonths:                doc.harvestMonths,
      seasonalWateringAdjustment:   doc.seasonalWateringAdjustment,
      minRainfallMm:                doc.minRainfallMm,
      waterLitersPerWatering:       doc.waterLitersPerWatering,
      auditHistory:                 doc.auditHistory.map(this._toAuditEntry),
      createdAt:            doc.createdAt,
      updatedAt:            doc.updatedAt,
      deletedAt:            doc.deletedAt,
    });
  }

  /**
   * Convierte una entidad PlantSpecies a documento MongoDB (sin _id).
   *
   * @param entity — Entidad PlantSpecies.
   * @returns Documento parcial para insertar/actualizar.
   */
  toDocument(entity: PlantSpecies): Omit<PlantSpeciesDocument, '_id'> {
    return {
      name:                 entity.name,
      scientificName:       entity.scientificName,
      image:                entity.image,
      careRequirements:     this._toCareRequirementsDoc(entity.careRequirements),
      climateCompatibility: entity.climateCompatibility,
      tips:                 entity.tips,
      createdBy:            entity.createdBy ? new ObjectId(entity.createdBy) : undefined,
      isPublic:             entity.isPublic,
      requiresPruning:      entity.requiresPruning,
      pruningMonths:         entity.pruningMonths,
      produceFruit:                 entity.produceFruit,
      harvestMonths:                entity.harvestMonths,
      seasonalWateringAdjustment:   entity.seasonalWateringAdjustment,
      minRainfallMm:                entity.minRainfallMm,
      waterLitersPerWatering:       entity.waterLitersPerWatering,
      auditHistory:                 entity.auditHistory.map(this._toAuditEntryDoc),
      createdAt:            entity.createdAt,
      updatedAt:            entity.updatedAt,
      deletedAt:            entity.deletedAt,
    };
  }

  /**
   * Convierte una entidad PlantSpecies al DTO de respuesta HTTP.
   *
   * @param entity — Entidad PlantSpecies.
   * @returns PlantSpeciesResponseDTO serializable.
   */
  toResponseDTO(entity: PlantSpecies): PlantSpeciesResponseDTO {
    return {
      id:                   entity.id,
      name:                 entity.name,
      scientificName:       entity.scientificName,
      image:                entity.image,
      careRequirements: {
        wateringDays:     entity.careRequirements.wateringDays,
        lightNeed:        entity.careRequirements.lightNeed,
        temperatureRange: entity.careRequirements.temperatureRange,
      },
      climateCompatibility: entity.climateCompatibility,
      tips:                 entity.tips,
      isPublic:             entity.isPublic,
      requiresPruning:              entity.requiresPruning,
      pruningMonths:                 entity.pruningMonths,
      produceFruit:                 entity.produceFruit,
      harvestMonths:                entity.harvestMonths,
      seasonalWateringAdjustment:   entity.seasonalWateringAdjustment,
      minRainfallMm:                entity.minRainfallMm,
      waterLitersPerWatering:       entity.waterLitersPerWatering,
    };
  }

  /**
   * Convierte subdocumento de cuidados a tipo de dominio.
   * @private
   */
  private _toCareRequirements(doc: CareRequirementsDocument): CareRequirements {
    return {
      wateringDays:     doc.wateringDays,
      lightNeed:        doc.lightNeed as LightNeed,
      temperatureRange: doc.temperatureRange,
    };
  }

  /**
   * Convierte cuidados de dominio a subdocumento MongoDB.
   * @private
   */
  private _toCareRequirementsDoc(req: CareRequirements): CareRequirementsDocument {
    return {
      wateringDays:     req.wateringDays,
      lightNeed:        req.lightNeed,
      temperatureRange: req.temperatureRange,
    };
  }

  /**
   * Convierte subdocumento de auditoría a tipo de dominio.
   * @private
   */
  private _toAuditEntry(doc: AuditEntryDocument): AuditEntry {
    return {
      date:    doc.date,
      userId:  doc.userId,
      changes: doc.changes,
    };
  }

  /**
   * Convierte entrada de auditoría de dominio a subdocumento MongoDB.
   * @private
   */
  private _toAuditEntryDoc(entry: AuditEntry): AuditEntryDocument {
    return {
      date:    entry.date,
      userId:  entry.userId,
      changes: entry.changes,
    };
  }
}
